Pinta.Core/Managers/ToolManager.cs

```diff
@@ -172,23 +185,32 @@ public void SetCurrentTool (Gdk.Key shortcut)
 			if (tool != null)
 				SetCurrentTool(tool);
 		}
-		
+
 		private BaseTool FindNextTool (Gdk.Key shortcut)
 		{
-			string key = shortcut.ToString ().ToUpperInvariant ();
-			
-			// Begin looking at the tool after the current one
-			for (int i = index + 1; i < Tools.Count; i++) {
-				if (Tools[i].ShortcutKey.ToString ().ToUpperInvariant () == key)
-					return Tools[i];
-			}
-				
-			// Begin at the beginning and look up to the current tool
-			for (int i = 0; i < index; i++) {
-				if (Tools[i].ShortcutKey.ToString ().ToUpperInvariant () == key)
-					return Tools[i];
+			foreach (var group in groupedTools)
```

Comment:
Can this loop be replaced with a direct lookup in the dictionary? (by making the dictionary always store the upper-case version of the key)

Reply:
I have been thinking about it. I put some changes in order to simplify this, but it is not ready yet. I misunderstood the task. And now I can say that tuple is unnecessary, so more refactoring needs to be done. It will take a day more.

Reply:
No problem! The changes so far look good

Reply:
I found a few bugs and fixed it. I hope now it works fine.